pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go

```diff
@@ -333,17 +334,32 @@ func (c *Reconciler) createTaskRun(logger *zap.SugaredLogger, rprt *resources.Re
 			},
 		},
 		Spec: v1alpha1.TaskRunSpec{
-			TaskRef: &v1alpha1.TaskRef{
-				Name: rprt.ResolvedTaskResources.TaskName,
-			},
 			Inputs: v1alpha1.TaskRunInputs{
 				Params: rprt.PipelineTask.Params,
 			},
 			ServiceAccount: sa,
-		},
-	}
+		}}
+
 	resources.WrapSteps(&tr.Spec, rprt.PipelineTask, rprt.ResolvedTaskResources.Inputs, rprt.ResolvedTaskResources.Outputs)
 
+	if pipelineTimeout != nil {
+		pTimeoutTime := pr.Status.StartTime.Add(pipelineTimeout.Duration)
+		if rprt.ResolvedTaskResources.TaskSpec.Timeout == nil || time.Now().Add(rprt.ResolvedTaskResources.TaskSpec.Timeout.Duration).After(pTimeoutTime) {
+			// Just in case something goes awry and we're creating the TaskRun after it should have already timed out,
+			// set a timeout of 0.
+			taskRunTimeout := pTimeoutTime.Sub(time.Now())
+			if taskRunTimeout < 0 {
+				taskRunTimeout = 0
+			}
+			rprt.ResolvedTaskResources.TaskSpec.Timeout = &metav1.Duration{Duration: taskRunTimeout}
+		}
+		tr.Spec.TaskSpec = rprt.ResolvedTaskResources.TaskSpec
+	} else {
```

Comment:
a bit confused over here, 
If timeout is defined, we set tr.Spec.TaskSpec in the else we set tr.Spec.TaskRef
Why so?

Reply:
Because I wanted to not be making changes to behavior except when necessary - I don’t like it much, though, and would be fine with just putting the TaskSpec there in all circumstances.

Reply:
hmm this is interesting, this actually has me leaning more toward what @shashwathi was saying about having the timeout on the Run rather than the spec.

if we change the timeout on the Task spec here, we're changing it for all future users of the Task as well which doesn't seem great

(i dont have any great suggestions but will try to give it a long think if no one comes up with anything in the meantime :D)

Reply:
I assumed that the `TaskSpec` here is just a copy of the `Task`'s `TaskSpec`', not a reference - am I mistaken or doing something wrong?

I think we really want to have the timeout be something you can specify on the `Task` - that's what our current API is (since `Task` has the field already), and it's definitely something you're going to want to be able to specify for a `Task` globally, not just for individual `TaskRun`s of that `Task`. It gets hairier here because we need to add or override the timeout for a `TaskRun` when the `Pipeline` timeout is specified. If this was the only thing we'd ever see a reason to need to override like this, then a new field on `TaskRunSpec` that lets you set an overriding timeout would feel most right to me, but I have a feeling we'll run into other cases like this in the future, at which point special-casing each of them feels unwieldy.

Reply:
Ah, I need to do a `DeepCopy` - fixing that, and then we can switch to the philosophical discussion. =)

Reply:
sorry if this was already covered in earlier discussions, but `TaskSpec` is optional in `TaskRun` and for the most case, `TaskRun` will just have a `TaskRef` to get info from the `Task` itself. Are you replacing the `TaskRef` when there is timeout?

Reply:
Yes, exactly. If there's a `Pipeline` timeout, the `TaskRun` gets a `TaskSpec` instead of a `TaskRef`.

Reply:
That's the part I don't like - sometimes there's a `TaskRef`, sometimes there's a `TaskSpec`...might it just make sense to always have the `TaskRun`s created by the `PipelineRun` reconciler contain a copy of the `TaskSpec`, even if we don't change anything in it?

Reply:
it depends on if we want to take a snapshot of the task at the time of creation of the taskRun (which I think makes sense) or if we care to pick up changes in a task

Reply:
Philosophically, I think that when you launch a `PipelineRun`, it should be using a snapshot of its `Pipeline` and `Task`s as of that moment - making changes to the underlying definitions behind a run in progress and having those changes get picked up *during* that run seems weird and wrong to me. But by that logic, we might want to actually be copying in the `TaskSpec` not just under all conditions but right when we create the `PipelineRun`, not just when we go to create the individual `TaskRun`s.

Reply:
yeah, I think we have an issue to take a snapshot of tasks at the start of a pipeline

Reply:
#279

Reply:
Ok, then I'm just going to change this logic here to always copy in the `TaskSpec`, regardless of whether there's a `Pipeline` timeout, and when #279 happens, that'll just change where we copy the `TaskSpec` from - i.e., change from copying it directly from the `Task` during `createTaskRun` to copying it from wherever we snapshot the `Task`s.